include/rsp.h

```diff
@@ -10,12 +10,206 @@
+ * 
+ * If you're using libdragon's build system (n64.mk), use DEFINE_RSP_UCODE()
+ * to initialize one of these.
+ */
+typedef struct {
+    uint8_t *code; void *code_end;
+    uint8_t *data; void *data_end;
+    const char *name; uint32_t start_pc;
+} rsp_ucode_t;
+
+/**
+ * @brief Define one RSP ucode compiled via libdragon's build system (n64.mk).
+ * 
+ * If you're using libdragon's build system (n64.mk), use DEFINE_RSP_UCODE() to
+ * define one ucode coming from a rsp_*.S file. For instance, if you wrote
+ * and compiled a ucode called rsp_math.S, you can use DEFINE_RSP_UCODE(rsp_math)
+ * to define it at the global level. You can then use rsp_load(&rsp_math) 
+ * to load it.
+ */
+#define DEFINE_RSP_UCODE(ucode_name) \
+    extern uint8_t ucode_name ## _text_start[]; \
+    extern uint8_t ucode_name ## _data_start[]; \
+    extern uint8_t ucode_name ## _text_end[0]; \
+    extern uint8_t ucode_name ## _data_end[0]; \
+    rsp_ucode_t ucode_name = (rsp_ucode_t){ \
+        .code = ucode_name ## _text_start, \
+        .data = ucode_name ## _data_start, \
+        .code_end = ucode_name ## _text_end, \
+        .data_end = ucode_name ## _data_end, \
+        .name = #ucode_name, .start_pc = 0, \
+    }
+
+/** @brief Initialize the RSP subsytem. */
+void rsp_init(void);
+
+/** @brief Load a RSP ucode.
+ * 
+ * This function allows to load a RSP ucode into the RSP internal memory.
+ * The function executes the transfer right away, so it is responsibility
+ * of the caller making sure that it's a good time to do it.
+ * 
+ * The function internally keeps a pointer to the last loaded ucode. If the
+ * ucode passed is the same, it does nothing. This makes it easier to write
+ * code that optimistically switches between different ucodes, but without
+ * forcing transfers every time.
+ * 
+ * @param[in]     ucode       Ucode to load into RSP
+ **/
+void rsp_load(rsp_ucode_t *ucode);
+
+/** @brief Run RSP ucode.
+ * 
+ * This function starts running the RSP, and wait until the ucode is finished.
+ */
+void rsp_run(void);
+
+/** @brief Run RSP async.
+ * 
+ * This function starts running the RSP in background. Use rsp_wait() to
+ * synchronize later.
+ */
+void rsp_run_async(void);
+
+/** @brief Wait until RSP is idle. */
+void rsp_wait(void);
+
+/** @brief Do a DMA transfer to load a piece of code into RSP IMEM.
+ * 
+ * This is a lower-level function that actually executes a DMA transfer
+ * from RDRAM to IMEM. Prefer using rsp_load instead.
+ * 
+ * @param[in]     code          Pointer to buffer in RDRAM containing code. 
+ *                              Must be aligned to 8 bytes.
+ * @param[in]     size          Size of the code to load. This is rounded up
+ *                              to 8 bytes.
+ * @param[in]     imem_offset   Byte offset in IMEM where to load the code.
+ *                              Must be a multiple of 8.
+ */
+void rsp_load_code(void* code, unsigned long size, unsigned int imem_offset);
```

Comment:
If a programmer was looking to do something like a microcode overlay, would using `imem_offset` be how they do it?

Reply:
Possibly, yes. In general I'm adding the offset because I feel like being limited to 0 is gratuitous, especially for loading/saving data to DMEM.